lib/BloqadeSchema/src/transform.jl

```diff
@@ -0,0 +1,389 @@
+    max_slope = device_capabilities.rydberg.global_value.detuningSlewRateMax
+    max_value = device_capabilities.rydberg.global_value.detuningMax
+    min_value = device_capabilities.rydberg.global_value.detuningMin
+
+
+    check_waveform(Δ,:Δ)
+    check_global(Δ,:Δ)
+    warn_duration(time_res,Δ,:Δ)
+    
+
+
+    if Δ isa PiecewiseLinearWaveform
+        Δt = piecewise_linear(;
+            clocks=set_resolution.(Δ.f.clocks,time_res),
+            values=set_resolution.(Δ.f.values,detune_res)
+        )
+        Δ_mask = (Δ=Δt,δ=nothing,Δi=1.0)
+        Δt = clip_waveform(Δt,:Δ,min_value,max_value)
+        return Δt,norm_diff_durations(Δ,Δt),Δ_mask
+
+    elseif Δ isa Waveform{F,T} where {F,T<:Real}
+        # arbitrary waveform must transform
+        
+        Δ_interp = piecewise_linear_interpolate(Δ,max_slope=max_slope,min_step=min_step,atol=0)
+        Δt = piecewise_linear(;
+            clocks=set_resolution.(Δ_interp.f.clocks,time_res),
+            values=set_resolution.(Δ_interp.f.values,detune_res)
+        )
+
+        Δt = clip_waveform(Δt,:Δ,min_value,max_value)
+        Δ_mask = (Δ=Δt,δ=nothing,Δi=1.0)
+
+        return Δt,norm_diff_durations(Δ,Δt),Δ_mask
+    elseif Δ isa Vector
+        throw(NotImplementedError("Not implemented."))
+
+        """
+        duration = Δ[1].duration
+        nsteps = Int((duration+duration%min_step)÷min_step)
+        clocks = collect(LinRange(0.0,duration,nsteps))
+        clocks = set_resolution.(clocks,time_res)
+
+        values = zeros(length(clocks),length(Δ))
+
+        @inbounds for (j,δ) in enumerate(Δ)
+            values[:,j] = δ.(clocks)
+        end
+
+        ((δ_values,Δi),(Δ_values,_)) = find_local_masks(values;name=:Δ)
+
+        Δ_values = set_resolution.(Δ_values,detune_res)
+        δ_values = set_resolution.(δ_values,common_detune_res)
+        Δi = set_resolution.(Δi,local_res)
+
+        Δ_mask = (
+            Δ=piecewise_linear(;clocks=clocks,values=Δ_values),
+            δ=piecewise_linear(;clocks=clocks,values=Δ_values),
+            Δi=Δi
+        )
+        Δt = [piecewise_linear(;clocks=clocks,values=Δ_values+δi.*δ_values) for δi in Δi]
+        error = [norm_diff_durations(δ,δt) for (δ,δt) in zip(Δ,Δt)]
+
+        return Δt,error,Δ_mask
+        """
+    end
+
+
+end
+
+function hardware_transform_parse(H::BloqadeExpr.RydbergHamiltonian,device_capabilities::DeviceCapabilities)
+    (atoms,ϕ,Ω,Δ) = get_rydberg_params(H)
+
+    ϕ,ϕ_error = hardware_transform_ϕ(ϕ,device_capabilities)
+    Ω,Ω_error = hardware_transform_Ω(Ω,device_capabilities)
+    Δ,Δ_error,Δ_mask = hardware_transform_Δ(Δ,device_capabilities)
+
+    pos_resolution = device_capabilities.lattice.geometry.positionResolution
+    new_atoms = [set_resolution.(pos,pos_resolution) for pos in atoms]
+
+    mse_atoms = sum(√sum((a .- b) .^ 2) for (a,b) in zip(new_atoms,atoms))/length(atoms)
+
+    info = (ϕ=ϕ_error,Ω=Ω_error,Δ=Δ_error,Δ_mask=Δ_mask,mse_atoms=mse_atoms)
```

Comment:
define a type to contain this using

```julia
Base.@kwdef struct XXXInfo
   ϕ_error
   # blabla
end
```

then you can just write

```julia
info = XXXInfo(;ϕ_error, Ω_error, #= blabla...=#)
```

Reply:
@johnzl-777